Pull based replication not support library/redis:2.8 in docker hub

The following is the log for replication job
```
2019-05-23T02:24:11Z [INFO] [/replication/transfer/image/transfer.go:95]: client for source registry [type: docker-hub, URL: https://hub.docker.com, insecure: false] created
2019-05-23T02:24:11Z [INFO] [/replication/transfer/image/transfer.go:105]: client for destination registry [type: harbor, URL: http://harbor-harbor-core, insecure: true] created
2019-05-23T02:24:11Z [INFO] [/replication/transfer/image/transfer.go:138]: copying library/redis:[2.8](source registry) to library/redis:[2.8](destination registry)...
2019-05-23T02:24:11Z [INFO] [/replication/transfer/image/transfer.go:157]: copying library/redis:2.8(source registry) to library/redis:2.8(destination registry)...
2019-05-23T02:24:11Z [INFO] [/replication/transfer/image/transfer.go:264]: pulling the manifest of image library/redis:2.8 ...
2019-05-23T02:24:12Z [INFO] [/replication/transfer/image/transfer.go:274]: the manifest of image library/redis:2.8 pulled
2019-05-23T02:24:12Z [ERROR] [/replication/transfer/image/transfer.go:224]: unsupported media type: application/octet-stream
2019-05-23T02:24:12Z [ERROR] [/replication/transfer/image/transfer.go:143]: unsupported media type: application/octet-stream
```

And the manifest for library/redis:2.8 in docker hub is below
```json
{
    "schemaVersion": 2,
    "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
    "config": {
        "mediaType": "application/octet-stream",
        "size": 5946,
        "digest": "sha256:481995377a044d40ca3358e4203fe95eca1d58b98a1d4c2d9cec51c0c4569613"
    },
    "layers": [
        {
            "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
            "size": 51356334,
            "digest": "sha256:51f5c6a04d83efd2d45c5fd59537218924bc46705e3de6ffc8bc07b51481610b"
        },
        {
            "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
            "size": 2035,
            "digest": "sha256:6c8ccd839b1db85aa4669156d14d3c09cb691d62e76d48b31b591403bdd43822"
        },
        {
            "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
            "size": 16629376,
            "digest": "sha256:0fded1c9651dac642c75ff5a09425f88832584cca23cdf3ff54cca655e6a9337"
        },
        {
            "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
            "size": 807927,
            "digest": "sha256:7f1aa6a73799308fbfc23049863ce6902092714c6ae07debc014c07f440f3661"
        },
        {
            "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
            "size": 2903046,
            "digest": "sha256:fbe8a4f1aa876020ae4aae0ff1bd9704d981714361cf85f773a8877e70de597c"
        },
        {
            "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
            "size": 96,
            "digest": "sha256:1a9852d2edd343460e89ebe0c4fa790063fe3c3e85c6618b45ddf127ef764372"
        },
        {
            "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
            "size": 395,
            "digest": "sha256:128182e1e85d1b381458ef6f98a78ab640080e72f5500a2a6fa25d8a6c766ef9"
        },
        {
            "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
            "size": 119,
            "digest": "sha256:b94de088b6d8ec958cd8ce3deb690e5c2a6f8c5b0f1368674234b5a44d709cbd"
        }
    ]
}
```

--- a/src/core/api/repository.go
+++ b/src/core/api/repository.go
@@ -25,6 +25,7 @@ import (
 	"time"
 
 	"errors"
+
 	"github.com/docker/distribution/manifest/schema1"
 	"github.com/docker/distribution/manifest/schema2"
 	"github.com/goharbor/harbor/src/common"
@@ -682,14 +683,17 @@ func getTagDetail(client *registry.Repository, tag string) (*tagDetail, error) {
 		Name: tag,
 	}
 
-	digest, _, payload, err := client.PullManifest(tag, []string{schema2.MediaTypeManifest})
+	digest, mediaType, payload, err := client.PullManifest(tag, []string{schema2.MediaTypeManifest})
 	if err != nil {
 		return detail, err
 	}
 	detail.Digest = digest
 
-	manifest := &schema2.DeserializedManifest{}
-	if err = manifest.UnmarshalJSON(payload); err != nil {
+	if strings.Contains(mediaType, "application/json") {
+		mediaType = schema1.MediaTypeManifest
+	}
+	manifest, _, err := registry.UnMarshal(mediaType, payload)
+	if err != nil {
 		return detail, err
 	}
 
@@ -699,7 +703,21 @@ func getTagDetail(client *registry.Repository, tag string) (*tagDetail, error) {
 		detail.Size += ref.Size
 	}
 
-	_, reader, err := client.PullBlob(manifest.Target().Digest.String())
+	// if the media type of the manifest isn't v2, doesn't parse image config
+	// and return directly
+	// this impacts that some detail information(os, arch, ...) of old images
+	// cannot be got
+	if mediaType != schema2.MediaTypeManifest {
+		log.Debugf("the media type of the manifest is %s, not v2, skip", mediaType)
+		return detail, nil
+	}
+	v2Manifest, ok := manifest.(*schema2.DeserializedManifest)
+	if !ok {
+		log.Debug("the manifest cannot be convert to DeserializedManifest, skip")
+		return detail, nil
+	}
+
+	_, reader, err := client.PullBlob(v2Manifest.Target().Digest.String())
 	if err != nil {
 		return detail, err
 	}

--- a/src/replication/transfer/image/transfer.go
+++ b/src/replication/transfer/image/transfer.go
@@ -211,18 +211,15 @@ func (t *transfer) copyContent(content distribution.Descriptor, srcRepo, dstRepo
 	case schema2.MediaTypeManifest:
 		// as using digest as the reference, so set the override to true directly
 		return t.copyImage(srcRepo, digest, dstRepo, digest, true)
-	// copy layer or image config
-	case schema2.MediaTypeLayer, schema2.MediaTypeImageConfig:
-		return t.copyBlob(srcRepo, dstRepo, digest)
 	// handle foreign layer
 	case schema2.MediaTypeForeignLayer:
 		t.logger.Infof("the layer %s is a foreign layer, skip", digest)
 		return nil
-	// others
+	// copy layer or image config
+	// the media type of the layer or config can be "application/octet-stream",
+	// schema1.MediaTypeManifestLayer, schema2.MediaTypeLayer, schema2.MediaTypeImageConfig
 	default:
-		err := fmt.Errorf("unsupported media type: %s", content.MediaType)
-		t.logger.Error(err.Error())
-		return err
+		return t.copyBlob(srcRepo, dstRepo, digest)
 	}
 }
 
@@ -288,6 +285,7 @@ func (t *transfer) handleManifest(manifest distribution.Manifest, repository, di
 	}
 	// manifest
 	if mediaType == schema1.MediaTypeManifest ||
+		mediaType == schema1.MediaTypeSignedManifest ||
 		mediaType == schema2.MediaTypeManifest {
 		return manifest, digest, nil
 	}

